nucliadb/src/nucliadb/search/search/chat/query.py

```diff
@@ -481,3 +518,371 @@ async def _prequery_find(prequery: PreQuery, index: int):
+    )
+
+
+async def find_retrieval(
+    kbid: str,
+    find_request: FindRequest,
+    x_ndb_client: NucliaDBClientType,
+    x_nucliadb_user: str,
+    x_forwarded_for: str,
+    metrics: Metrics,
+) -> tuple[KnowledgeboxFindResults, bool, Fetcher, Reranker]:
+    # TODO: Remove once the feature has been fully rolled out
+    if not has_feature(const.Features.ASK_DECOUPLED, context={"kbid": kbid}):
+        results, incomplete, parsed = await find(
+            kbid,
+            find_request,
+            x_ndb_client,
+            x_nucliadb_user,
+            x_forwarded_for,
+            metrics=metrics,
+        )
+        # this has already been asserted inside the find() call
+        assert parsed.retrieval.reranker is not None, "find parser must provide a reranking algorithm"
+        reranker = get_reranker(parsed.retrieval.reranker)
+        return results, incomplete, parsed.fetcher, reranker
+
+    return await rao_find(
+        kbid,
+        find_request,
+        x_ndb_client,
+        x_nucliadb_user,
+        x_forwarded_for,
+        metrics=metrics,
+    )
+
+
+async def rao_find(
+    kbid: str,
+    find_request: FindRequest,
+    x_ndb_client: NucliaDBClientType,
+    x_nucliadb_user: str,
+    x_forwarded_for: str,
+    metrics: Metrics,
+) -> tuple[KnowledgeboxFindResults, bool, Fetcher, Reranker]:
+    """
+    Calls to NucliaDB retrieve and augment primitives to perform the text block search.
+    It returns the results as KnowledgeboxFindResults to comply with the existing find
+    interface (/ask logic is tightly coupled with /find).
+
+    # retrieve
+    # augment
+    # rerank
+    # convert results to KnowledgeboxFindResults
+    """
+    fetcher, retrieval_request, reranker = await rao_parse_find(kbid, find_request)
+
+    query = find_request.query
+    rephrased_query = None
+    if retrieval_request.query.keyword:
+        if find_request.query != retrieval_request.query.keyword.query:
+            rephrased_query = retrieval_request.query.keyword.query
+
+    retrieval_response = await retrieve(kbid, retrieval_request)
+    matches = retrieval_response.matches
+
+    relations = None
+    if FindOptions.RELATIONS in find_request.features:
+        # the user asked for a legacy relations search, as we don't support it
+        # in the /retrieve endpoint but we must maintain bw/c with /find
+        # responses, we call it with to get just this part of the response
+        #
+        # TODO: replace this with an sdk call
+        find_response, _, _ = await find(
+            kbid,
+            FindRequest(
+                features=[FindOptions.RELATIONS],
+                # needed for automatic entity detection
+                query=query,
+                # used for "hardcoded" graph queries
+                query_entities=find_request.query_entities,
+            ),
+            x_ndb_client,
+            x_nucliadb_user,
+            x_forwarded_for,
+            metrics,
+        )
```

Comment:
Can't we use the graph endpoints instead via sdk?

Reply:
We don't expose the same in the graph endpoints, we should tweak them and to another model conversion in RAO before having the proper one. I don't think using /find for this specific case is that bad. If we deprecate/remove the relations feature from the /ask endpoint we won't need this anymore